pkg/util/kubernetes/autoscalers/datadogclient.go

```diff
@@ -0,0 +1,234 @@
+				retryInterval:      minRetryInterval,
+			},
+		},
+	}
+	for _, endpoint := range endpoints {
+		client := datadog.NewClient(endpoint.APIKey, endpoint.APPKey)
+		client.HttpClient.Transport = httputils.CreateHTTPTransport()
+		client.RetryTimeout = 3 * time.Second
+		client.ExtraHeader["User-Agent"] = "Datadog-Cluster-Agent"
+		client.SetBaseUrl(endpoint.URL)
+		ddFallbackClient.clients = append(
+			ddFallbackClient.clients,
+			&datadogIndividualClient{
+				client:             client,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			})
+	}
+
+	return ddFallbackClient, nil
+}
+
+func (ic *datadogIndividualClient) queryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	// TODO: check for err 500 (server side) instead of any kind of error because switching to DR doesn’t make sense if that’s the query which is bad.
+	if series, err := ic.client.QueryMetrics(from, to, query); err == nil {
+		ic.lastQuerySucceeded = true
+		ic.lastSuccess = time.Now()
+		ic.retryInterval /= 2
+		if ic.retryInterval < minRetryInterval {
+			ic.retryInterval = minRetryInterval
+		}
+		return series, err
+	} else { // revive:disable-line:indent-error-flow
+		ic.lastQuerySucceeded = false
+		ic.lastFailure = time.Now()
+		ic.retryInterval *= 2
+		if ic.retryInterval > maxRetryInterval {
+			ic.retryInterval = maxRetryInterval
+		}
+		return series, err
+	}
+}
+
+func (ic *datadogIndividualClient) hasFailedRecently() bool {
+	if ic.lastQuerySucceeded {
+		return false
+	}
+
+	return ic.lastFailure.Add(ic.retryInterval).After(time.Now())
+}
+
+func (cl *datadogFallbackClient) QueryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	errs := errors.New("Failed to query metrics on all endpoints")
+
+	skippedClients := []*datadogIndividualClient{}
+
+	for i, c := range cl.clients {
+
+		if c.hasFailedRecently() {
+			skippedClients = append(skippedClients, c)
+			continue
+		}
+
+		if series, err := c.queryMetrics(from, to, query); err == nil {
+			if i != cl.lastUsedClient {
+				log.Warnf("Switching external metrics source provider from %s to %s",
+					cl.clients[cl.lastUsedClient].client.GetBaseUrl(),
+					c.client.GetBaseUrl())
+			}
+			cl.lastUsedClient = i
+			return series, nil
+		} else { // revive:disable-line:indent-error-flow
+			log.Infof("Failed to query metrics on %s: %#v", c.client.GetBaseUrl(), err)
+			errs = fmt.Errorf("%w, Failed to query metrics on %s: %v", errs, c.client.GetBaseUrl(), err)
+		}
+	}
+
+	for _, c := range skippedClients {
+		log.Infof("Although we shouldn’t query %s because of the backoff policy, we’re going to do so because no other valid endpoint succeeded so far.", c.client.GetBaseUrl())
+		if series, err := c.queryMetrics(from, to, query); err == nil {
+			return series, nil
+		} else { // revive:disable-line:indent-error-flow
+			errs = fmt.Errorf("%w, Failed to query metrics on %s: %v", errs, c.client.GetBaseUrl(), err)
+		}
+	}
+
+	return nil, errs
```

Comment:
I think the error message here should mention that we failed to reach all endpoints. And the caller simply logs the message again, so maybe we don't need to duplicate the error messages? WDYT?

Reply:
Yes, this error mentions that all endpoints failed: https://github.com/DataDog/datadog-agent/pull/9908/files#diff-dd0473b11ce864431885968451bae7c5732b13ac050abee682b3571130dc6c32R165

The thing is that, if not all endpoints failed, it’s still valuable to have logs for individual endpoint failures.

Reply:
> The thing is that, if not all endpoints failed, it’s still valuable to have logs for individual endpoint failures.

Yes, I wasn't suggesting we remove them. Just that it might be helpful for troubleshooting to say here something along the lines: `fmt.Errorf("All endpoints failed: %w", errs)`. `errs` looked a bit redundant to me in the beginning, but it might be actually better this way.